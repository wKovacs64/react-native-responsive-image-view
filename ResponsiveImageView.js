import React, { Component } from 'react';
import { Image } from 'react-native';
import PropTypes from 'prop-types';
import resolveAssetSource from 'react-native/Libraries/Image/resolveAssetSource';

class ResponsiveImageView extends Component {
  initialState = {
    loading: true,
    error: '',
    aspectRatio: undefined,
  };

  state = this.initialState;

  componentDidMount() {
    const { source } = this.props;
    if (source.uri) {
      // Retrieve image dimensions from URI
      this.pendingGetImageSize = this.getImageSize(
        source.uri,
        this.handleImageSizeSuccess,
        this.handleImageSizeFailure,
      );
    } else {
      // Retrieve image dimensions from imported resource
      const imageSource = resolveAssetSource(source);
      if (imageSource && imageSource.width && imageSource.height) {
        const { width, height } = imageSource;
        this.handleImageSizeSuccess(width, height);
      } else {
        this.handleImageSizeFailure(
          new Error('Failed to retrieve image dimensions.'),
        );
      }
    }
  }

  componentWillUnmount() {
    // istanbul ignore else: noop
    if (this.pendingGetImageSize) {
      this.pendingGetImageSize.cancel();
    }
  }

  // A cancelable version of Image.getSize, adapted from
  // https://github.com/kodefox/react-native-flex-image
  getImageSize = (uri, onImageSizeSuccess, onImageSizeFailure) => {
    let totallyCanceled = false;

    Image.getSize(
      uri,
      (width, height) => {
        if (!totallyCanceled) {
          onImageSizeSuccess(width, height);
        }
      },
      error => {
        if (!totallyCanceled) {
          onImageSizeFailure(error);
        }
      },
    );

    return {
      cancel: () => {
        totallyCanceled = true;
      },
    };
  };

  getImageProps = ({ source, style = {}, ...props } = {}) => ({
    source: this.props.source,
    style: [style, { height: '100%', width: '100%' }],
    ...props,
  });

  getViewProps = ({ style = {}, ...props } = {}) => ({
    style: [style, { aspectRatio: this.state.aspectRatio }],
    ...props,
  });

  handleImageSizeSuccess = (width, height) => {
    const { aspectRatio, onLoad } = this.props;
    this.setState(
      {
        ...this.initialState,
        loading: false,
        aspectRatio: aspectRatio || width / height,
      },
      () => onLoad(),
    );
  };

  handleImageSizeFailure = err => {
    const { onError } = this.props;
    this.setState(
      {
        ...this.initialState,
        loading: false,
        error: err.message,
      },
      () => onError(this.state.error),
    );
  };

  render() {
    const { component: ComponentOrFunction, render, children } = this.props;
    const { loading, error } = this.state;
    const bag = {
      loading,
      error,
      getImageProps: this.getImageProps,
      getViewProps: this.getViewProps,
    };

    // component injection
    if (ComponentOrFunction) {
      // class component
      if (ComponentOrFunction.prototype.render) {
        return <ComponentOrFunction {...bag} />;
      }

      // stateless functional component (SFC)
      return ComponentOrFunction(bag);
    }

    // render prop
    if (render) {
      return render(bag);
    }

    // function-as-children
    if (typeof children === 'function') {
      return children(bag);
    }

    // no renderer provided, but children exist - just render the children as-is
    if (children && React.Children.count(children) > 0) {
      return React.Children.only(children);
    }

    return null;
  }
}

ResponsiveImageView.displayName = 'ResponsiveImageView';

ResponsiveImageView.propTypes = {
  aspectRatio: PropTypes.number,
  component: PropTypes.func,
  render: PropTypes.func,
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  onError: PropTypes.func,
  onLoad: PropTypes.func,
  source: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      uri: PropTypes.string.isRequired,
    }),
  ]).isRequired,
};

ResponsiveImageView.defaultProps = {
  aspectRatio: undefined,
  component: undefined,
  render: undefined,
  children: undefined,
  onError: () => {},
  onLoad: () => {},
};

export default ResponsiveImageView;
