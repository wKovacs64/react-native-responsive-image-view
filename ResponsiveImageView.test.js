import React from 'react';
import { Text, View } from 'react-native';
import { shallow } from 'enzyme';
import ResponsiveImageView from './ResponsiveImageView';

// Test data
const mockUriGood = 'mockUriGood';
const mockUriBad = 'mockUriBad';
const mockUriSlowGood = 'mockUriSlowGood';
const mockUriSlowBad = 'mockUriSlowBad';
const mockResourceGood = 1000;
const mockResourceBad = 9999;
const mockWidth = 800;
const mockHeight = 600;
const aspectRatio = 16 / 9;
const consumerViewProps = {
  hitSlop: {
    top: 10,
    bottom: 10,
    left: 0,
    right: 0,
  },
  style: {
    padding: 20,
  },
};
const consumerImageProps = {
  onLayout: jest.fn().mockName('Image#onLayout'),
  style: {
    height: '50%',
    width: '50%',
  },
};

// Mocks
jest.mock('Image', () => ({
  getSize(uri, onLoad, onError) {
    switch (uri) {
      case mockUriGood:
        onLoad(mockWidth, mockHeight);
        break;
      case mockUriBad:
        onError(new Error(uri));
        break;
      case mockUriSlowGood:
        setImmediate(onLoad);
        break;
      case mockUriSlowBad:
        setImmediate(() => {
          onError(uri);
        });
        break;
      default:
        throw new Error(`Unexepcted URI value in test: ${uri}`);
    }
  },
}));
jest.mock('react-native/Libraries/Image/resolveAssetSource', () => res => {
  if (res === mockResourceGood) {
    return { width: mockWidth, height: mockHeight };
  }
  return null;
});

// Tests
describe('rendering order (component > render > FAC > children > null)', () => {
  it('should render component if provided', () => {
    const MyRenderComponent = props => <View {...props} />;
    const render = jest.fn();
    const children = jest.fn();
    const wrapper = shallow(
      <ResponsiveImageView
        source={{ uri: mockUriGood }}
        component={MyRenderComponent}
        render={render}
      >
        {children}
      </ResponsiveImageView>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(render.mock.calls).toMatchSnapshot('render');
    expect(children.mock.calls).toMatchSnapshot('children');
  });

  it('should call render if no component was provided', () => {
    const render = jest.fn();
    const children = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriGood }} render={render}>
        {children}
      </ResponsiveImageView>,
    );
    expect(render.mock.calls).toMatchSnapshot('render');
    expect(children.mock.calls).toMatchSnapshot('children');
  });

  it('should call children function if no component or render prop was provided', () => {
    const children = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriGood }}>
        {children}
      </ResponsiveImageView>,
    );
    expect(children.mock.calls).toMatchSnapshot('children');
  });

  it('should render children if no component, render prop, or FAC was provided', () => {
    expect(
      shallow(
        <ResponsiveImageView source={{ uri: mockUriGood }}>
          <View>
            <Text>Hello from non-functional children!</Text>
          </View>
        </ResponsiveImageView>,
      ),
    ).toMatchSnapshot();
  });

  it('should not render anything if no renderer was provided', () => {
    expect(
      shallow(<ResponsiveImageView source={{ uri: mockUriGood }} />).type(),
    ).toBeNull();
  });
});

describe('component injection', () => {
  it('should render the component with expected parameters on success', () => {
    const MyRenderComponent = props => <View {...props} />;
    expect(
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          component={MyRenderComponent}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should render the component with expected parameters on failure', () => {
    const MyRenderComponent = props => <View {...props} />;
    expect(
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriBad }}
          component={MyRenderComponent}
        />,
      ),
    ).toMatchSnapshot();
  });

  it('should render class components', () => {
    // eslint-disable-next-line react/prefer-stateless-function
    class MyRenderClassComponent extends React.Component {
      render() {
        return <View {...this.props} />;
      }
    }
    expect(
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          component={MyRenderClassComponent}
        />,
      ),
    ).toMatchSnapshot();
  });
});

describe('render prop', () => {
  it('should call render with expected parameters on success', () => {
    const render = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriGood }} render={render} />,
    );
    expect(render.mock.calls).toMatchSnapshot();
  });

  it('should call render with expected parameters on failure', () => {
    const render = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriBad }} render={render} />,
    );
    expect(render.mock.calls).toMatchSnapshot();
  });
});

describe('function-as-children', () => {
  it('should call children with expected parameters on success', () => {
    const children = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriGood }}>
        {children}
      </ResponsiveImageView>,
    );
    expect(children.mock.calls).toMatchSnapshot();
  });

  it('should call children with expected parameters on failure', () => {
    const children = jest.fn();
    shallow(
      <ResponsiveImageView source={{ uri: mockUriBad }}>
        {children}
      </ResponsiveImageView>,
    );
    expect(children.mock.calls).toMatchSnapshot();
  });
});

describe('getViewProps', () => {
  it('should include consumer provided aspectRatio', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          aspectRatio={aspectRatio}
          source={{ uri: mockUriGood }}
          onError={reject}
          onLoad={resolve}
          render={({ getViewProps }) => {
            expect(getViewProps()).toMatchSnapshot();
          }}
        />,
      );
    }));

  it('should calculate aspectRatio if not provided', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          onError={reject}
          onLoad={resolve}
          render={({ getViewProps }) => {
            expect(getViewProps()).toMatchSnapshot();
          }}
        />,
      );
    }));

  it('should compose consumer props with own props', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          onError={reject}
          onLoad={resolve}
          render={({ getViewProps }) => {
            expect(getViewProps({ ...consumerViewProps })).toMatchSnapshot();
          }}
        />,
      );
    }));
});

describe('getImageProps', () => {
  it('should set height and width to 100%', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          onError={reject}
          onLoad={resolve}
          render={({ getImageProps }) => {
            expect(getImageProps()).toMatchSnapshot();
          }}
        />,
      );
    }));

  it('should compose consumer props with own props', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          onError={reject}
          onLoad={resolve}
          render={({ getImageProps }) => {
            expect(getImageProps({ ...consumerImageProps })).toMatchSnapshot();
          }}
        />,
      );
    }));
});

describe('completion callbacks', () => {
  it('should call provided onLoad on URI success', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriGood }}
          onLoad={resolve}
          onError={reject}
        />,
      );
    }));

  it('should call provided onError on URI failure', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={{ uri: mockUriBad }}
          onLoad={reject}
          onError={resolve}
        />,
      );
    }));

  it('should call provided onLoad on imported resource success', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={mockResourceGood}
          onLoad={resolve}
          onError={reject}
        />,
      );
    }));

  it('should call provided onError on imported resource failure', () =>
    new Promise((resolve, reject) => {
      shallow(
        <ResponsiveImageView
          source={mockResourceBad}
          onLoad={reject}
          onError={resolve}
        />,
      );
    }));

  it('should not call onLoad on success after unmounting', () =>
    new Promise((resolve, reject) => {
      const wrapper = shallow(
        <ResponsiveImageView
          source={{ uri: mockUriSlowGood }}
          onLoad={reject}
          onError={reject}
        />,
      );
      wrapper.unmount();
      setImmediate(resolve);
    }));

  it('should not call onError on failure after unmounting', () =>
    new Promise((resolve, reject) => {
      const wrapper = shallow(
        <ResponsiveImageView
          source={{ uri: mockUriSlowBad }}
          onLoad={reject}
          onError={reject}
        />,
      );
      wrapper.unmount();
      setImmediate(resolve);
    }));
});
