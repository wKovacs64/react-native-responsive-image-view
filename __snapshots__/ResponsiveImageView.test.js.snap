// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component injection should render class components 1`] = `
<MyRenderClassComponent
  error=""
  getImageProps={[Function]}
  getViewProps={[Function]}
  loading={false}
/>
`;

exports[`component injection should render the component with expected parameters on failure 1`] = `
<View
  error="mockUriBad"
  getImageProps={[Function]}
  getViewProps={[Function]}
  loading={false}
/>
`;

exports[`component injection should render the component with expected parameters on success 1`] = `
<View
  error=""
  getImageProps={[Function]}
  getViewProps={[Function]}
  loading={false}
/>
`;

exports[`function-as-children should call children with expected parameters on failure 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "mockUriBad",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`function-as-children should call children with expected parameters on success 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`getImageProps should compose consumer props with own props 1`] = `
Object {
  "onLayout": [MockFunction Image#onLayout],
  "source": Object {
    "uri": "mockUriGood",
  },
  "style": Array [
    Object {
      "height": "50%",
      "width": "50%",
    },
    Object {
      "height": "100%",
      "width": "100%",
    },
  ],
}
`;

exports[`getImageProps should compose consumer props with own props 2`] = `
Object {
  "onLayout": [MockFunction Image#onLayout],
  "source": Object {
    "uri": "mockUriGood",
  },
  "style": Array [
    Object {
      "height": "50%",
      "width": "50%",
    },
    Object {
      "height": "100%",
      "width": "100%",
    },
  ],
}
`;

exports[`getImageProps should set height and width to 100% 1`] = `
Object {
  "source": Object {
    "uri": "mockUriGood",
  },
  "style": Array [
    Object {},
    Object {
      "height": "100%",
      "width": "100%",
    },
  ],
}
`;

exports[`getImageProps should set height and width to 100% 2`] = `
Object {
  "source": Object {
    "uri": "mockUriGood",
  },
  "style": Array [
    Object {},
    Object {
      "height": "100%",
      "width": "100%",
    },
  ],
}
`;

exports[`getViewProps should calculate aspectRatio if not provided 1`] = `
Object {
  "style": Array [
    Object {},
    Object {
      "aspectRatio": undefined,
    },
  ],
}
`;

exports[`getViewProps should calculate aspectRatio if not provided 2`] = `
Object {
  "style": Array [
    Object {},
    Object {
      "aspectRatio": 1.3333333333333333,
    },
  ],
}
`;

exports[`getViewProps should compose consumer props with own props 1`] = `
Object {
  "hitSlop": Object {
    "bottom": 10,
    "left": 0,
    "right": 0,
    "top": 10,
  },
  "style": Array [
    Object {
      "padding": 20,
    },
    Object {
      "aspectRatio": undefined,
    },
  ],
}
`;

exports[`getViewProps should compose consumer props with own props 2`] = `
Object {
  "hitSlop": Object {
    "bottom": 10,
    "left": 0,
    "right": 0,
    "top": 10,
  },
  "style": Array [
    Object {
      "padding": 20,
    },
    Object {
      "aspectRatio": 1.3333333333333333,
    },
  ],
}
`;

exports[`getViewProps should include consumer provided aspectRatio 1`] = `
Object {
  "style": Array [
    Object {},
    Object {
      "aspectRatio": undefined,
    },
  ],
}
`;

exports[`getViewProps should include consumer provided aspectRatio 2`] = `
Object {
  "style": Array [
    Object {},
    Object {
      "aspectRatio": 1.7777777777777777,
    },
  ],
}
`;

exports[`render prop should call render with expected parameters on failure 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "mockUriBad",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`render prop should call render with expected parameters on success 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`rendering order (component > render > FAC > children > null) should call children function if no component or render prop was provided: children 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`rendering order (component > render > FAC > children > null) should call render if no component was provided: children 1`] = `Array []`;

exports[`rendering order (component > render > FAC > children > null) should call render if no component was provided: render 1`] = `
Array [
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": true,
    },
  ],
  Array [
    Object {
      "error": "",
      "getImageProps": [Function],
      "getViewProps": [Function],
      "loading": false,
    },
  ],
]
`;

exports[`rendering order (component > render > FAC > children > null) should render children if no component, render prop, or FAC was provided 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Hello from non-functional children!
  </Text>
</View>
`;

exports[`rendering order (component > render > FAC > children > null) should render component if provided 1`] = `
<View
  error=""
  getImageProps={[Function]}
  getViewProps={[Function]}
  loading={false}
/>
`;

exports[`rendering order (component > render > FAC > children > null) should render component if provided: children 1`] = `Array []`;

exports[`rendering order (component > render > FAC > children > null) should render component if provided: render 1`] = `Array []`;
